--- a/config/BUILDCONFIG.gn
+++ b/config/BUILDCONFIG.gn
@@ -291,6 +291,8 @@
   assert(
       false,
       "emscripten is not a supported target_os. It is available only as secondary toolchain.")
+} else if (target_os == "wasi") {
+  _default_toolchain = "//build/toolchain/wasi:wasm32"
 } else {
   assert(false, "Unsupported target_os: $target_os")
 }
@@ -327,12 +329,13 @@
 is_ios = current_os == "ios"
 is_linux = current_os == "linux"
 is_mac = current_os == "mac"
-is_wasm = current_os == "emscripten"
+is_wasm = current_os == "emscripten" || current_os == "wasi"
+is_wasi = current_os == "wasi"
 is_watchos = current_os == "watchos"
 is_win = current_os == "win" || current_os == "winuwp"

 is_apple = is_ios || is_mac || is_watchos
-is_posix = !is_win && !is_fuchsia
+is_posix = !is_win && !is_fuchsia && !is_wasi

 # =============================================================================
 # TARGET DEFAULTS
--- a/config/compiler/BUILD.gn
+++ b/config/compiler/BUILD.gn
@@ -356,6 +356,8 @@
     configs += [ "//build/config/aix:compiler" ]
   } else if (current_os == "zos") {
     configs += [ "//build/config/zos:compiler" ]
+  } else if (current_os == "wasi") {
+    configs += [ "//build/config/wasi:compiler" ]
   }

   configs += [
@@ -456,7 +458,7 @@
       } else if (is_chromeos) {
         cflags += [ "-fstack-protector-strong" ]
       } else if (is_posix || is_fuchsia) {
-        if (current_os != "aix") {
+        if (current_os != "aix" && current_os != "wasi") {
           # Not available on aix.
           cflags += [ "-fstack-protector" ]
         }
@@ -3185,8 +3187,12 @@
     # All configs using /DEBUG should include this:
     configs += [ ":win_pdbaltpath" ]
   } else if (is_wasm) {
-    cflags = [ "-gseparate-dwarf" ]
-    ldflags = [ "-gseparate-dwarf" ]
+    if (!is_wasi) {
+      cflags = [ "-gseparate-dwarf" ]
+      ldflags = [ "-gseparate-dwarf" ]
+    } else {
+      cflags = []
+    }
   } else {
     if (use_debug_fission) {
       configs += [ ":split_dwarf" ]
